### ISSUE TITLE
Define expected sbatch script for parallel execution

### ISSUE THREAD

**Issue by tcompa** [fractal-analytics-platform/fractal-server#572]
Let's try to build a minimal set of requirements for the sbatch scripts that will be needed in #493.

---

For these examples, we prepare a `task.py` script which runs for a small time and simply writes its start/end time stamps:
```python
import time
import sys
import logging

component = sys.argv[1]           # This what we want to parallelize over
sleep_time = float(sys.argv[2])   # How long to sleep

logging.basicConfig(format="%(asctime)s; %(message)s")
logging.warning(f"{component=}, {sleep_time=}; START")
time.sleep(sleep_time)
logging.warning(f"{component=}, {sleep_time=}; END")
```

---

Then we want to write a sbatch script that runs `task.py` for N values of `component`. As a first step, we will include all `N` execution commands in the same sbatch script, but we want to be able to control the maximum number of `task.py` simultaneous executions - let's call this `max_tasks`.
Expected behavior (say that `sleep_time` is set to 1 second):
* If `max_tasks=N`, then the `N` tasks run in parallel, that is, the total runtime is ~1 second.
* If `max_tasks=N/2`, then the runtime is ~2 seconds.
* ...
* If `max_tasks=1`, then the `N` tasks run serially, that is, the total runtime is `N` seconds.

This is the very first requirement for the sbatch jobs that will be needed in #493. We can use this issue to log ongoing tests.

**Comment by tcompa** [comment]
In the previous example, what is actually providing the value of `max_tasks` is not `--ntasks=3`, as one would guess, but rather the ratio between `--mem=3000MB` and `MEM_PER_TASK=1000MB`, which is equal to 3.

To see this, we can leave `ntasks=3` and try to execute more tasks:
```
#!/bin/bash
#SBATCH --partition=main
#SBATCH --err=err_1
#SBATCH --out=out_1
#SBATCH --cpus-per-task=1
#SBATCH --ntasks=3
#SBATCH --mem=5000MB       # This cannot be smaller than ntasks*MEM_PER_TASK

MEM_PER_TASK=1000MB
PYTHON=/data/homes/fractal/anaconda3/bin/python
srun --ntasks=1 --cpus-per-task=$SLURM_CPUS_PER_TASK --mem=$MEM_PER_TASK $PYTHON task.py a 1 &
srun --ntasks=1 --cpus-per-task=$SLURM_CPUS_PER_TASK --mem=$MEM_PER_TASK $PYTHON task.py b 1 &
srun --ntasks=1 --cpus-per-task=$SLURM_CPUS_PER_TASK --mem=$MEM_PER_TASK $PYTHON task.py c 1 &
srun --ntasks=1 --cpus-per-task=$SLURM_CPUS_PER_TASK --mem=$MEM_PER_TASK $PYTHON task.py c 1 &
srun --ntasks=1 --cpus-per-task=$SLURM_CPUS_PER_TASK --mem=$MEM_PER_TASK $PYTHON task.py c 1 &
wait
```
which results in all tasks being executed simultaneously:
```
2023-03-16 12:58:31,915; component='c', sleep_time=1.0; START
2023-03-16 12:58:32,027; component='a', sleep_time=1.0; START
2023-03-16 12:58:32,212; component='c', sleep_time=1.0; START
2023-03-16 12:58:32,318; component='c', sleep_time=1.0; START
2023-03-16 12:58:32,422; component='b', sleep_time=1.0; START
2023-03-16 12:58:32,916; component='c', sleep_time=1.0; END
2023-03-16 12:58:33,028; component='a', sleep_time=1.0; END
2023-03-16 12:58:33,213; component='c', sleep_time=1.0; END
2023-03-16 12:58:33,319; component='c', sleep_time=1.0; END
2023-03-16 12:58:33,425; component='b', sleep_time=1.0; END
```

**Comment by tcompa** [comment]
For the record, the examples above were wrong (likely they were apparently working due to some specific configuration of the docker-slurm cluster). The latest commit (https://github.com/fractal-analytics-platform/fractal-server/commit/1d25e22970fc7419dcb1c1d68369ccf76bd80844) goes much closer to the expected behavior.

**Comment by tcompa** [comment]
> We DO NOT test anything related to resources - for the moment. We only ask for a single CPU per task, and we have no control of what kind of resources (memory/CPU) will be requested from SLURM. This implies that, for the moment we cannot say where one of our job will end up in the existing SLURM queue - and whether this plan would work when scaling up the number of Fractal tasks per submission script.

Small addition: we do not have any check in place for whether multithreading is taking place correctly, within this configuration. We may try to do so by creating a heavy (numpy-based) test task, and by looking at thread use during its execution - but it won't be easy (or reasonably possible?) to catch this through a test.

**Comment by tcompa** [comment]
Current status (as of 66eeded1): in principle we can also correctly set the required memory per-job, as is visible in this squeue output (obtained for n_ftasks_tot=10, n_ftasks_per_script=4, n_parallel_ftasks_per_script=3):
```
   JOBID PARTITION           NAME    STATE MIN_CPU CPUS MIN_MEMORY        TIME   TIME_LIMIT    NODES  NODELIST(REASON)
      25      main submit_batch_2 COMPLETI       1    1       600M        0:01    UNLIMITED        1  slurmnode1
      23      main submit_batch_0  RUNNING       1    1       600M        0:02    UNLIMITED        1  slurmnode1
      24      main submit_batch_1  RUNNING       1    1       600M        0:02    UNLIMITED        1  slurmnode1
```
but it is still not clear whether we are setting the CPU requirements correctly.

**Comment by tcompa** [comment]
> but it is still not clear whether we are setting the CPU requirements correctly.

Useful examples:
* https://support.ceci-hpc.be/doc/_contents/SubmittingJobs/SlurmFAQ.html#q05-how-do-i-create-a-parallel-environment

**Comment by tcompa** [comment]
> For my understanding, what happens when we do: ...

Good question!
Slightly different from my first guess, your example still works as it should. That is, when submitting
```bash
#!/bin/bash
#SBATCH --partition=main
#SBATCH --err=err
#SBATCH --out=out
#SBATCH --cpus-per-task=2
#SBATCH --ntasks=2
#SBATCH --mem=400MB


COMMAND="/data/homes/fractal/anaconda3/bin/python fake_task_for_timing.py"
srun --ntasks=1 --cpus-per-task=$SLURM_CPUS_PER_TASK --mem=200MB $COMMAND 0 1.0 logs &
srun --ntasks=1 --cpus-per-task=$SLURM_CPUS_PER_TASK --mem=200MB $COMMAND 1 1.0 logs &
wait

srun --ntasks=1 --cpus-per-task=$SLURM_CPUS_PER_TASK --mem=200MB $COMMAND 2 1.0 logs &
srun --ntasks=1 --cpus-per-task=$SLURM_CPUS_PER_TASK --mem=200MB $COMMAND 3 1.0 logs &
wait

srun --ntasks=1 --cpus-per-task=$SLURM_CPUS_PER_TASK --mem=200MB $COMMAND 4 1.0 logs &
wait
```
all tasks are run a single time -- see logs:
```
2023-03-21 07:31:56.534503, START for component='0' and sleep_time=1.0
2023-03-21 07:31:56.645820, START for component='1' and sleep_time=1.0
2023-03-21 07:31:57.535614, END   for component='0' and sleep_time=1.0
2023-03-21 07:31:57.646925, END   for component='1' and sleep_time=1.0
2023-03-21 07:31:57.906264, START for component='2' and sleep_time=1.0
2023-03-21 07:31:58.039488, START for component='3' and sleep_time=1.0
2023-03-21 07:31:58.907354, END   for component='2' and sleep_time=1.0
2023-03-21 07:31:59.040603, END   for component='3' and sleep_time=1.0
2023-03-21 07:31:59.282951, START for component='4' and sleep_time=1.0
2023-03-21 07:32:00.284057, END   for component='4' and sleep_time=1.0
```

**Comment by tcompa** [comment]
FYI, the "wrong" behavior you were referring to (some tasks being executed multiple times) can be reproduced and then fixed as in the following examples.

TL;DR setting `--ntasks=1` in the `srun` command seem to remove the multiple-executions risk for a given task.

### "Wrong" example (multiple executions)
Script
```bash
#!/bin/bash
#SBATCH --partition=main
#SBATCH --err=err
#SBATCH --out=out
#SBATCH --ntasks=2

COMMAND="/data/homes/fractal/anaconda3/bin/python fake_task_for_timing.py"
srun $COMMAND 0 1.0 logs
```
Logs
```
2023-03-21 07:37:35.205838, START for component='0' and sleep_time=1.0
2023-03-21 07:37:35.214324, START for component='0' and sleep_time=1.0
2023-03-21 07:37:36.206980, END   for component='0' and sleep_time=1.0
2023-03-21 07:37:36.215437, END   for component='0' and sleep_time=1.0
```


### "Correct" example
Script
```bash
#!/bin/bash
#SBATCH --partition=main
#SBATCH --err=err
#SBATCH --out=out
#SBATCH --ntasks=2

COMMAND="/data/homes/fractal/anaconda3/bin/python fake_task_for_timing.py"
srun --ntasks=1 $COMMAND 0 1.0 logs
```
Logs
```
2023-03-21 07:38:14.145971, START for component='0' and sleep_time=1.0
2023-03-21 07:38:15.147095, END   for component='0' and sleep_time=1.0
```

**Comment by tcompa** [comment]
One more example (thanks @enricotagliavini).

(side-note: I now included a bit of randomness in the python task, so that tasks can finish at less predictable times)

This submission script
```bash
#!/bin/bash
#SBATCH --partition=main
#SBATCH --err=err
#SBATCH --out=out
#SBATCH --cpus-per-task=2
#SBATCH --ntasks=2
#SBATCH --mem=400MB

COMMAND="/data/homes/fractal/anaconda3/bin/python fake_task_for_timing.py"
OPTIONS="1 2 3 4 5 6 7 8"
for OPT in $OPTIONS; do
    srun --ntasks=1 --cpus-per-task=$SLURM_CPUS_PER_TASK --mem=200MB $COMMAND $OPT 1.0 logs &
done
wait
```
leads to this log
```
2023-03-21 12:46:53.848273, START for component='1' and sleep_time=7.0
2023-03-21 12:46:53.969630, START for component='8' and sleep_time=8.0
2023-03-21 12:47:00.855378, END   for component='1' and sleep_time=7.0
srun: Job step creation temporarily disabled, retrying
srun: Job step creation temporarily disabled, retrying
srun: Job step creation temporarily disabled, retrying
srun: Job step creation temporarily disabled, retrying
srun: Job step creation temporarily disabled, retrying
2023-03-21 12:47:01.157134, START for component='2' and sleep_time=6.0
2023-03-21 12:47:01.977737, END   for component='8' and sleep_time=8.0
srun: Job step created
2023-03-21 12:47:02.232205, START for component='6' and sleep_time=8.0
2023-03-21 12:47:07.163246, END   for component='2' and sleep_time=6.0
srun: Job step created
2023-03-21 12:47:07.400217, START for component='7' and sleep_time=10.0
2023-03-21 12:47:10.240344, END   for component='6' and sleep_time=8.0
srun: Job step created
2023-03-21 12:47:10.538412, START for component='5' and sleep_time=2.0
2023-03-21 12:47:12.540518, END   for component='5' and sleep_time=2.0
srun: Job step created
2023-03-21 12:47:12.794218, START for component='4' and sleep_time=5.0
2023-03-21 12:47:17.410334, END   for component='7' and sleep_time=10.0
srun: Job step created
2023-03-21 12:47:17.639208, START for component='3' and sleep_time=1.0
2023-03-21 12:47:17.799332, END   for component='4' and sleep_time=5.0
2023-03-21 12:47:18.640305, END   for component='3' and sleep_time=1.0
```

Apart from the "srun: Job step creation temporarily disabled, retrying" messages, the task execution is the one we would expect - with a task starting as soon as one "slot" becomes available.

**Comment by enricotagliavini** [comment]
You can try adding the ```--quiet``` option to see if it helps with the clutter in the output. It might get rid of the ```srun: Job step creation temporarily disabled, retrying``` lines
